be/src/formats/parquet/predicate_filter_evaluator.h

```diff
@@ -104,19 +179,63 @@ struct PredicateFilterEvaluator {
         }
 
         for (const auto& child : node.compound_children()) {
-            ASSIGN_OR_RETURN(auto cur_row_ranges_opt, child.visit(*this));
+            ASSIGN_OR_RETURN(auto cur_row_ranges_opt, child.visit(*this, Evaluator::BLOOM_FILTER));
             if (cur_row_ranges_opt.has_value()) {
                 PredicateFilterEvaluatorUtils::merge_row_ranges<Type>(row_ranges, cur_row_ranges_opt.value());
             }
         }
         return row_ranges;
     }
 
+    template <CompoundNodeType Type>
+    StatusOr<std::optional<SparseRange<uint64_t>>> operator()(const PredicateCompoundNode<Type>& node) {
+        return this->operator()(node, Evaluator::ALL);
+    }
+
+    template <CompoundNodeType Type>
+    StatusOr<std::optional<SparseRange<uint64_t>>> operator()(const PredicateCompoundNode<Type>& node, Evaluator mode) {
+        DCHECK(group_reader != nullptr);
+        bool row_group_filtered = false;
+        std::optional<SparseRange<uint64_t>> row_ranges;
+        if (mode & Evaluator::STATS_ZONEMAP) {
+            auto ret = visit_for_zonemap(node);
+            if (ret.ok() && ret.value().has_value()) {
+                row_group_filtered = ret.value()->span_size() == 0;
+                row_ranges = std::move(ret.value());
+            }
+        }
+        if (mode & Evaluator::PAGE_INDEX) {
+            if (enable_page_index && !row_group_filtered) {
+                auto ret = visit_for_page_index(node);
+                if (ret.ok() && ret.value().has_value()) {
+                    row_group_filtered = ret.value()->span_size() == 0;
+                    row_ranges = std::move(ret.value());
+                }
+            }
+        }
+        if (mode & Evaluator::BLOOM_FILTER) {
+            if (enable_bloom_filter && !row_group_filtered) {
+                if (mode != Evaluator::BLOOM_FILTER) {
+                    row_ranges_before_bf = row_ranges;
+                }
+                auto ret = visit_for_bloom_filter(node);
+                if (ret.ok() && ret.value().has_value()) {
+                    row_group_filtered = ret.value()->span_size() == 0;
+                    if (row_group_filtered) {
+                        row_ranges = std::move(ret.value());
```

Comment:
If mode = all, will the subsequent range overwrite the result calculated previously?